src/lib.rs

```diff
@@ -1,25 +1,45 @@
-use std::io::BufRead;
+use std::path::Path;
 
+use rayon::iter::ParallelBridge;
+use rayon::prelude::*;
+
+mod chunks;
 mod counter;
 mod key_finder;
 
+pub use chunks::chunks;
 pub use counter::{Counter, KeyCount};
 pub use key_finder::KeyFinder;
 
-pub fn top_few_from_stream<T: BufRead>(
-    reader: T,
+fn sum(mut l: Counter, r: Counter) -> Counter {
+    for (key, count) in r.counts().into_iter() {
+        l.add(&key, count);
+    }
+    l
+}
+
+pub fn top_few_from_stream(
+    path: &Path,
     kf: &KeyFinder,
     num: usize,
-) -> Result<Vec<KeyCount>, anyhow::Error> {
-    let mut counter = Counter::new(num);
-    let mut s = String::new();
-    for ln in reader.lines() {
-        s.clear();
-        let ln = ln?;
-        if let Ok(key) = kf.key(&ln, &mut s) {
-            counter.add(key)
-        }
-    }
+) -> anyhow::Result<Vec<KeyCount>> {
+    let total = chunks(path, 1 << 26)?
+        .collect::<Vec<_>>()
+        .into_iter()
+        .par_bridge()
+        .map(|reader| {
+            let mut counter = Counter::new(0);
```

Comment:
Why did you use `0` as the `Counter::new()` argument here (and below in the `fold()` call)? It seems to me that it would be better to always use the cutoff.

Reply:
When you map over the lines you don't need to calculate top at all. It's only relevant to update counts.

Only once you fold it makes sense to keep track of top - even then some top computation are discarded.

Reply:
In that case, what if we make `num` an `Option`? In that case we can also avoid the unnecessary `top` updates.

Reply:
Ok make sense to use Option to make this explicit.